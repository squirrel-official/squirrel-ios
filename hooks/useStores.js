/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
import { createContext, useContext } from 'react';

import RootStore from '../stores/RootStore';

export const storesContext = createContext({
	rootStore: new RootStore()
});

export const useStores = () => useContext(storesContext);
