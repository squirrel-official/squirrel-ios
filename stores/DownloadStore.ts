/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import { action, computed, decorate, observable } from 'mobx';
import { format } from 'mobx-sync-lite';

import DownloadModel from '../models/DownloadModel';

export const DESERIALIZER = (data: unknown) => {
	const deserialized = new Map<string, DownloadModel>();
	Object.entries(data).forEach(([ key, dl ]) => {
		const model = new DownloadModel(
			dl.itemId,
			dl.serverId,
			dl.serverUrl,
			dl.apiKey,
			dl.title,
			dl.filename,
			dl.downloadUrl
		);
		model.isComplete = dl.isComplete;
		// isDownloading is ignored
		model.isNew = dl.isNew;
		deserialized.set(key, model);
	});
	return deserialized;
};

export default class DownloadStore {
	downloads = new Map<string, DownloadModel>();

	get newDownloadCount() {
		return Array.from(this.downloads.values())
			.filter(d => d.isNew)
			.length;
	}

	add(download: DownloadModel) {
		// Do not allow duplicate downloads
		if (!this.downloads.has(download.key)) {
			this.downloads.set(download.key, download);
		}
	}

	reset() {
		this.downloads = new Map();
	}
}

decorate(DownloadStore, {
	downloads: [
		format(DESERIALIZER),
		observable
	],
	newDownloadCount: computed,
	add: action,
	reset: action
});
