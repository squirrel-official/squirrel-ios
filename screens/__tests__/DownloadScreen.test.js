/**
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

import { NavigationContainer } from '@react-navigation/native';
import { render } from '@testing-library/react-native';
import React from 'react';

import { useStores } from '../../hooks/useStores';
import DownloadModel from '../../models/DownloadModel';
import DownloadStore from '../../stores/DownloadStore';
import DownloadScreen from '../DownloadScreen';

const mockSetOptions = jest.fn();
jest.mock('@react-navigation/native', () => {
	const actualNav = jest.requireActual('@react-navigation/native');
	return {
		...actualNav,
		useNavigation: () => ({
			setOptions: mockSetOptions
		})
	};
});

const mockDownloadStore = new DownloadStore();
jest.mock('../../hooks/useStores');
useStores.mockImplementation(() => ({
	rootStore: {
		downloadStore: mockDownloadStore
	}
}));

describe('DownloadScreen', () => {
	it('should render correctly', () => {
		mockDownloadStore.add(new DownloadModel(
			'item-id',
			'server-id',
			'https://example.com/',
			'api-key',
			'title',
			'file name.mkv',
			'https://example.com/download'
		));
		mockDownloadStore.add(new DownloadModel(
			'item-id-2',
			'server-id',
			'https://test2.example.com/',
			'api-key',
			'other title',
			'other file name.mkv',
			'https://test2.example.com/download'
		));

		const { toJSON } = render(
			<NavigationContainer>
				<DownloadScreen />
			</NavigationContainer>
		);

		expect(toJSON()).toMatchSnapshot();
		expect(mockSetOptions).toHaveBeenCalled();
	});
});
