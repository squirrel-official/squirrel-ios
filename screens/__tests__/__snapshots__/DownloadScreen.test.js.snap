// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadScreen should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "right",
      "left",
    ]
  }
  style={
    Object {
      "backgroundColor": undefined,
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "marginTop": 1,
      }
    }
    data={
      Array [
        Object {
          "__mobx_sync_versions__": undefined,
          "apiKey": "api-key",
          "downloadUrl": "https://example.com/download",
          "filename": "file name.mkv",
          "isComplete": false,
          "isNew": true,
          "itemId": "item-id",
          "serverId": "server-id",
          "serverUrl": "https://example.com/",
          "sessionId": "uuid-0",
          "title": "title",
        },
        Object {
          "__mobx_sync_versions__": undefined,
          "apiKey": "api-key",
          "downloadUrl": "https://test2.example.com/download",
          "filename": "other file name.mkv",
          "isComplete": false,
          "isNew": true,
          "itemId": "item-id-2",
          "serverId": "server-id",
          "serverUrl": "https://test2.example.com/",
          "sessionId": "uuid-1",
          "title": "other title",
        },
      ]
    }
    extraData={
      Object {
        "server-id_item-id": [Circular],
        "server-id_item-id-2": [Circular],
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "borderTopWidth": 0.5,
                "flexDirection": "row",
                "padding": 14,
              }
            }
            testID="padView"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                Object {
                  "colors": Object {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": Object {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 17,
                  }
                }
                testID="title"
              >
                title
              </Text>
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 15,
                  }
                }
                testID="subtitle"
              >
                file name.mp4
              </Text>
            </View>
            <View
              style={
                Object {
                  "paddingLeft": 16,
                }
              }
            />
            <ActivityIndicator
              testID="loading-indicator"
            />
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View>
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderBottomWidth": 0.5,
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
            testID="padView"
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "flex-start",
                    "flex": 1,
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
              theme={
                Object {
                  "colors": Object {
                    "black": "#242424",
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "grey": "rgba(0, 0, 0, 0.54)",
                        "primary": "#2196f3",
                        "searchBg": "#dcdce1",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "default": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "grey": "#7d7d7d",
                        "primary": "#007aff",
                        "searchBg": "#dcdce1",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                      "web": Object {
                        "error": "#ff190c",
                        "grey": "#393e42",
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#ca71eb",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#ca71eb",
                    "success": "#52c41a",
                    "warning": "#faad14",
                    "white": "#ffffff",
                  },
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 17,
                  }
                }
                testID="title"
              >
                other title
              </Text>
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#242424",
                    "fontSize": 15,
                  }
                }
                testID="subtitle"
              >
                other file name.mp4
              </Text>
            </View>
            <View
              style={
                Object {
                  "paddingLeft": 16,
                }
              }
            />
            <ActivityIndicator
              testID="loading-indicator"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RNCSafeAreaView>
`;
